### PR TITLE
genvimvim.lua: fix matching functions

### PR DESCRIPTION
**Pull request by blueyed**
Fixes https://github.com/neovim/neovim/issues/5060.

### CODE REVIEW

File: src/nvim/eval.c [hunk 1]
```diff
@@ -6679,7 +6679,9 @@ static struct fst {
   char f_min_argc;              /* minimal number of arguments */
   char f_max_argc;              /* maximal number of arguments */
   void        (*f_func)(typval_T *args, typval_T *rvar);
-  /* implementation of function */
+  /* Implementation of function.
+   * The lines below are matched/used by scripts/gen_events.lua.
+   */
```
Comment (blueyed) [inline src/nvim/eval.c]: Added a comment to hopefully avoid issues for this in the future.
But we might also add some assert in genvimvim.lua instead?

Comment (blueyed) [inline src/nvim/eval.c]: Did so.  Now it also matches functions containing underscores..

**Comment by mhinz** [comment]
Reverts of fixups.. could you rebase, please? :)

**Comment by blueyed** [comment]
@mhinz 
Just use the Github squash-merge button - or is it not enabled for this repo?

**Comment by blueyed** [comment]
Amended.

**Comment by bfredl** [comment]
Not sure what this fixes, but I wonder if #4934 fixes this or has the same problem.

**Comment by mhinz** [comment]
@blueyed Merging is not the problem, reviewing is.

I'll wait for others to chime in, since I have no idea about that part of the code.

CC @ZyX-I

**Comment by ZyX-I** [comment]
@bfredl That and #3922 (#4934 has commits from there) should fix this problem.

@mhinz `assert()` must not be there, “did not find a function in line” is completely valid condition for e.g. comments or `#if`s.

**Comment by blueyed** [comment]
> @blueyed Merging is not the problem, reviewing is.

In the "Files changed" tab you would only see the resulting diff.

@bfredl 
From quickly looking at https://github.com/neovim/neovim/pull/4934/files#diff-02dea228e0e60a44b1abde65aae4a22cL116 it might fix it as well.

**Comment by blueyed** [comment]
@bfredl 
To verify, check if `vimFuncName` is assigned to `bufnr` in `call bufnr('%')`.

**Comment by bfredl** [comment]
It sounds like it wouldn't be too hard to write a test that fails on master and this fixes.

**Comment by ZyX-I** [comment]
This PR only needs to change one line:

>     local func_name = line:match('^  {"(%w+)",')

to

>     local func_name = line:match('^  { "([%w_]+)",')

. Test would be simple:

```Patch
diff --git a/.ci/common/test.sh b/.ci/common/test.sh
index 225d88e..18c643d 100644
--- a/.ci/common/test.sh
+++ b/.ci/common/test.sh
@@ -83,8 +83,9 @@ install_nvim() {
     exit 1
   }
 
+  local genvimsynf=syntax/vim/generated.vim
   # Check that all runtime files were installed
-  for file in doc/tags syntax/vim/generated.vim $(
+  for file in doc/tags $genvimsynf $(
     cd runtime ; git ls-files | grep -e '.vim$' -e '.ps$' -e '.dict$' -e '.py$' -e '.tutor$'
   ) ; do
     if ! test -e "${INSTALL_PREFIX}/share/nvim/runtime/$file" ; then
@@ -93,6 +94,13 @@ install_nvim() {
     fi
   done
 
+  # Check that generated syntax file has function names
+  local gpat='syn keyword vimFuncName .*eval'
+  if cat $genvimsynf | grep -e "$gpat" ; then
+    echo "It appears that $genvimsynf does not contain $gpat"
+    exit 1
+  fi
+
   for file in $(
     cd runtime ; git ls-files | grep -e '.awk$' -e '.sh$' -e '.bat$'
   ) ; do
```

**Comment by blueyed** [comment]
@ZyX-I 
Well, it's the basic change of this PR, but it also adds an `assert` (which I think is a good test - since the build will fail there), and then removes the unnecessary `if`.

**Comment by blueyed** [comment]
But i can see that we want to have #4934 instead.
My comment to @bfredl was about what this PR fixes after all.
I could imagine that #4934 could use some sort of `assert` also and fail in case there are no functions found during build already?!

**Comment by blueyed** [comment]
I have not checked #4934, but when it takes some more time to be merged we might get this fixed through this PR first?!

**Comment by ZyX-I** [comment]
@blueyed No, it is not. `// deprecated, replaced by api_nvim_foo\n { "foo", … }` is completely valid code, but your `assert` rejects it. `#ifdef`ing functions is also completely valid. `assert()` must not be there.

**Comment by ZyX-I** [comment]
@blueyed Nobody here said that you should wait until #4934.

**Comment by ZyX-I** [comment]
#4934 needs no assert on not found functions, test I posted above is enough.

**Comment by bfredl** [comment]
I would imagine a screen test, displaying with `ft=vim`
```
call curbuf('%')
call notafunction()
```
to test the entire information flow.

**Comment by blueyed** [comment]
@ZyX-I 
Yes, any comment etc might be valid code - but it is working currently like that.
The assertion could be changed to only check for "at least one function" of course etc.

I agreed that a screen test for this would be best.
But on the other hand it's still good to fail during the build / parsing of the function file already for now.

**Comment by ZyX-I** [comment]
@blueyed Suggested test will fail when ability to install Neovim is tested. If you want you may add additional assert regarding “at least one function”, but I see no sense in this because no functions will fail with my test later and your assert will be removed in any case. The assert that is there currently must not be there, I do not know when #4934 will be merged and it is not impossible that some ported Vim patch will add `#if` or a comment.

**Comment by blueyed** [comment]
Ok.  I'll add your test then to this PR.

**Comment by blueyed** [comment]
Done, slighty adjusted.

**Comment by ZyX-I** [comment]
LGTM

**Comment by blueyed** [comment]
Please squash-merge it then, or let me know if I should update the branch for the PR.

**Comment by justinmk** [comment]
@blueyed The builds are failing:

```
grep: syntax/vim/generated.vim: No such file or directory
It appears that syntax/vim/generated.vim does not contain syn keyword vimFuncName .*eval.
```

File: .ci/common/test.sh [hunk 2]
```diff
@@ -93,6 +94,13 @@ install_nvim() {
     fi
   done
 
+  # Check that generated syntax file has function names, #5060.
+  local gpat='syn keyword vimFuncName .*eval'
+  if ! grep -q "$gpat" $genvimsynf ; then
```
Comment (justinmk) [inline .ci/common/test.sh]: line 89 has `cd runtime` so I guess `"runtime/$genvimsynf"` is needed here.

File: .ci/common/test.sh [hunk 3]
```diff
@@ -83,8 +83,9 @@ install_nvim() {
     exit 1
   }
 
+  local genvimsynf=syntax/vim/generated.vim
   # Check that all runtime files were installed
-  for file in doc/tags syntax/vim/generated.vim $(
+  for file in doc/tags $genvimsynf $(
     cd runtime ; git ls-files | grep -e '.vim$' -e '.ps$' -e '.dict$' -e '.py$' -e '.tutor$'
```
Comment (justinmk) [inline .ci/common/test.sh]: It seems like `doc/tags $genvimsynf ` should be `runtime/doc/tags "runtime/$genvimsynf"`

**Comment by blueyed** [comment]
The gcc-4.9 job on MacOS (Travis) timed out:

> No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

**Comment by ZyX-I** [comment]
Now there are two additional `fixup!` commits in history.

**Comment by mhinz** [comment]
(That's why I dislike fixup commits in PRs. They're super useful for local editing, but should be out-rebased when pushing to the PR branch.)

**Comment by justinmk** [comment]
It's my fault. I did not know it would bother anyone.

**Comment by blueyed** [comment]
Yeah.  I usually clean it up after the review is done, but for PRs that result in a single commit this can be done from the GitHub UI nowadays, too.

@mhinz 
I think fixup-commits are useful in PRs, too - since you can see what got changed over time..

**Pull request merged**